daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -124,62 +118,68 @@ final class ImmArray[+A] private (
     * but rather returning an empty array. If you want Seq's behavior,
     * use `relaxedSlice`.
     */
-  def strictSlice(from: Int, until: Int): ImmArray[A] = {
-    if (from < 0 || from >= length || until < 0 || until > length) {
+  def strictSlice(from: Int, until: Int): ImmArray[A] =
+    if (from < 0 || from >= length || until < 0 || until > length)
       throw new IndexOutOfBoundsException(
         s"strictSlice arguments out of bounds for ImmArray. length: $length, from: $from, until: $until")
-    }
-
-    relaxedSlice(from, until)
-  }
+    else
+      relaxedSlice(from, until)
 
   /** O(1), like `strictSlice`, but does not crash on out of bounds errors -- follows the same semantics
     * as Seq's slice.
     */
   def relaxedSlice(from0: Int, until0: Int): ImmArray[A] = {
-    val from = Math.max(Math.min(from0, length), 0)
-    val until = Math.max(Math.min(until0, length), 0)
-
+    val from = (from0 min length) max 0
+    val until = (until0 min length) max 0
     val newLen = until - from
-    if (newLen <= 0) {
-      ImmArray.empty[A]
-    } else {
+    if (isEmpty)
+      ImmArray.empty
+    else
       new ImmArray(start + from, newLen, array)
-    }
   }
 
   /** O(n) */
-  def map[B](f: A => B): ImmArray[B] = {
-    val newArray: mutable.ArraySeq[B] = new mutable.ArraySeq(length)
-    for (i <- indices) {
-      newArray(i) = f(uncheckedGet(i))
+  def map[B](f: A => B): ImmArray[B] =
+    if (isEmpty)
+      ImmArray.empty
+    else {
+      val newArray = new Array[AnyRef](length)
+      var i = 0
+      var j = start
+      while (i < length) {
+        newArray(i) = f(uncheckedGet(i)).asInstanceOf[AnyRef]
```

Comment:
I assume you meant this since the `j` wouldn't make too much sense otherwise:
```suggestion
        newArray(i) = f(array(j)).asInstanceOf[AnyRef]
```

Reply:
good catch.